test/jdk/java/awt/event/MouseEvent/AWTPanelSmoothWheel.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2009, 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.awt.BorderLayout;
+import java.awt.Color;
+import java.awt.Frame;
+import java.awt.Panel;
+
+/*
+ * @test
+ * @bug 6730447
+ * @summary [Win] To verify the support for high resolution mouse wheel on Windows.
+ *          AWT panel needs to support high-res mouse wheel rotation.
+ * @requires (os.family == "windows")
+ * @library /java/awt/regtesthelpers
+ * @build PassFailJFrame
+ * @run main/manual AWTPanelSmoothWheel
+ */
+
+public class AWTPanelSmoothWheel {
+    public static final String INSTRUCTIONS = """
+            This test is relevant for windows platforms and mouses with high-resolution wheel,
+            please just press pass if this is not the case.
+
+            Place the mouse cursor above the green panel and rotate the mouse wheel,
+            the test will print all mouse wheel event messages into the logging panel
+            below the instruction window.
+            Please make sure that some of the messages have non-zero 'wheelRotation' value,
+            and also check if the test works OK if the mouse wheel is rotated very slow.
```

Comment:
Sounds good, test can be made semi-automated test.
How long or many MouseWheelEvents do we check ? Probably 10-15 would be a good number?

Reply:
@aivanov-jdk Semi-automated the test, it now passes when 5 or more  MouseWheekEvent of type: scrollType=WHEEL_UNIT_SCROLL & wheelRotation != 0 is recorded. Updated the instructions accordingly.

Reply:
I think we should update the instructions.

A high resolution mouse produces events like this:
```text
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(164,90),absolute(1135,365),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=0,preciseWheelRotation=0.25] on panel0
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(164,90),absolute(1135,365),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=0,preciseWheelRotation=0.25] on panel0
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(164,90),absolute(1135,365),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=0,preciseWheelRotation=0.25] on panel0
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(133,46),absolute(1104,321),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=1,preciseWheelRotation=1.0] on panel0
```

Thus, `wheelRotation=0` while `preciseWheelRotation=0.25`. When it reaches, `preciseWheelRotation=1.0`, the event contains **`wheelRotation=1`**.

A regular mouse with mouse-wheel notches produces the following events:
```text
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(90,115),absolute(1061,390),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=1,preciseWheelRotation=1.0] on panel0
java.awt.event.MouseWheelEvent[MOUSE_WHEEL,(176,76),absolute(1147,351),button=0,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=3,wheelRotation=1,preciseWheelRotation=1.0] on panel0
```

As you can see, `scrollType=WHEEL_UNIT_SCROLL` is the same in both cases. The only difference is that high-resolution mouse produces events where **`preciseWheelRotation < 1.0`**.

This means that you should also track the value of `preciseWheelRotation`. Otherwise, the test will pass when a regular mouse is used.

You may show a warning to the user if you detect, let's say, 5 events with `wheelRotation=1` but none of the events had `wheelRotation=0` and `preciseWheelRotation < 1.0`.

Reply:
I believe since there a lot of conditions that needs to be checked manually, it is better to convert this test from semi-automatic to completely manual.

I can parse the MouseWheelEvent to display only the info needed by the user such as PreciseWheelRotation and WheelRotation for easier checking and include a warning when not using a hi-res mouse.

Reply:
@aivanov-jdk Thanks for reviewing this in-depth and pointing out the difference between wheelRotation & preciseWheelRotation when using hi-res mouse.

Reply:
@aivanov-jdk Since preciseWheelRotation amount depends on user's movement of the high-res mouse and there might be a need to look through logs for few events, it is better to convert this test from semi-automated to manual. Does this sound good. Updated the PR with the latest changes.

Reply:
@honkar-jdk I think the test could still remain semi-automatic. Can the test verify the conditions? I see no reason why it can't… yet it requires more code. It can be postponed to a later time, if necessary.

I don't think we should make the tester read the log and verify the conditions which can be verified automatically.